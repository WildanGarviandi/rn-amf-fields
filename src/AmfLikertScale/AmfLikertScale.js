import React, { Component } from 'react'
import PropTypes from 'prop-types'
import fieldPropTypes from '../fieldPropTypes'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { interpolateRdYlGn } from 'd3-scale-chromatic'

class AmfLikertScale extends Component {
  constructor(props) {
    super(props)

    this.state = {
      circleDiameter: 0,
      itemMargin: 0,
      valid: false
    }
  }

  componentDidMount() {
    // this.validate()
  }

  componentDidUpdate(prevProps) {
    // if (prevProps.selectedIndex !== this.props.selectedIndex) this.validate()
  }

  validate = () => {
    const { required, pristine, selectedIndex } = this.props

    if (!required) {
      this.setState({valid: true})
      return
    }

    if (pristine) {
      this.setState({valid: true})
      return
    }

    if (selectedIndex !== -1) {
      this.setState({valid: true})
      return
    }

    this.setState({valid: false})
    return
  }

  renderScale(item, index) {
    let interpolation = 0
    if (index < this.props.options.length/2) interpolation = index/this.props.options.length
    else interpolation = (index+1)/this.props.options.length

    return (
      <TouchableOpacity
        key={index}
        style={[
          style.scaleItem,
          {
            borderColor: interpolateRdYlGn(interpolation),
            backgroundColor: this.props.value == item.value ? interpolateRdYlGn(interpolation) : 'white',
            width: this.state.circleDiameter,
            height: this.state.circleDiameter,
            borderRadius: this.state.circleDiameter/2
          }
        ]}
        onPress={() => {
          this.props.onChange(item.value)
        }}
      >
        <Text style={[style.scaleText, {color: this.props.value == item.value ? 'white' : interpolateRdYlGn(interpolation), fontSize: this.state.circleDiameter/2.5}]}>{item.value}</Text>
      </TouchableOpacity>
    )
  }

  render() {
    let validationStyle = {}
    let validationText
    if (!this.state.valid) {
      validationStyle = { borderColor: 'red', borderWidth: 1 }
      validationText = <Text style={style.errorText}>{this.props.label + ' harus dipilih'}</Text>
    }

    return (
      <View style={{marginVertical: 8}}>
        <Text style={style.label}>{this.props.label}</Text>
        <View style={[{padding: 8}, validationStyle]}>
          <View style={style.scaleContainer}
            onLayout={(event) => {
              let base = event.nativeEvent.layout.width/this.props.options.length
              let itemMargin = base/this.props.options.length*1.5
              let diameter = (base)-(itemMargin)
              this.setState({circleDiameter: diameter, itemMargin: itemMargin})
            }}
          >
            { this.props.options.map(this.renderScale.bind(this)) }
          </View>
          <View style={[style.noteContainer, {marginHorizontal: this.state.itemMargin/2}]}>
            <Text style={{textAlign: 'center', width: this.state.circleDiameter}}>{this.props.options[0].label || 'Sangat Tidak Setuju'}</Text>
            <Text style={{textAlign: 'center', width: this.state.circleDiameter}}>{this.props.options[this.props.options.length-1].label || 'Sangat Setuju'}</Text>
          </View>
        </View>
        { validationText }
      </View>
    )
  }
}

const style = StyleSheet.create({
  errorText: {
    color: 'red',
    margin: 4,
    marginBottom: 0
  },
  label: {
    color: '#454545',
    marginRight: 16,
    marginBottom: 8
  },
  scaleContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    marginBottom: 16
  },
  noteContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  scaleItem: {
    borderWidth: 1,
    justifyContent: 'center'
  },
  scaleText: {
    textAlign: 'center'
  }
})

const itemShape = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired
})

AmfLikertScale.propTypes = fieldPropTypes

export default AmfLikertScale
