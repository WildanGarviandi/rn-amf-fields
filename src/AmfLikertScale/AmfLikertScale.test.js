import React from 'react'
import AmfLikertScale from './AmfLikertScale'
import renderer from 'react-test-renderer'

test('AmfLikertScale without validation', () => {
  let props = {
    label: 'Label',
    onPress: () => {},
    items: [
      {label:'Sangat Setuju', value: 1},
      {label:'Setuju Setuju', value: 2},
      {label:'Tida Setuju', value: 3},
      {label:'Sangat Tidak Setuju', value: 4},
    ],
    value: null
  }

  const component = renderer.create(<AmfLikertScale {...props} />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

