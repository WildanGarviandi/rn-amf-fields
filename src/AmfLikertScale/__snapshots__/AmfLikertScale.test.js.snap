// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmfLikertScale without validation 1`] = `
<View
  style={
    Object {
      "marginVertical": 8,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#454545",
        "marginBottom": 8,
        "marginRight": 16,
      }
    }
  >
    Label
  </Text>
  <View
    style={
      Array [
        Object {
          "padding": 8,
        },
        Object {},
      ]
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginBottom": 16,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "rgb(165, 0, 38)",
            "borderRadius": 0,
            "borderWidth": 1,
            "height": 0,
            "justifyContent": "center",
            "opacity": 1,
            "width": 0,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "rgb(165, 0, 38)",
                "fontSize": 0,
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "rgb(248, 141, 82)",
            "borderRadius": 0,
            "borderWidth": 1,
            "height": 0,
            "justifyContent": "center",
            "opacity": 1,
            "width": 0,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "rgb(248, 141, 82)",
                "fontSize": 0,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "rgb(133, 203, 103)",
            "borderRadius": 0,
            "borderWidth": 1,
            "height": 0,
            "justifyContent": "center",
            "opacity": 1,
            "width": 0,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "rgb(133, 203, 103)",
                "fontSize": 0,
              },
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "rgb(0, 104, 55)",
            "borderRadius": 0,
            "borderWidth": 1,
            "height": 0,
            "justifyContent": "center",
            "opacity": 1,
            "width": 0,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "rgb(0, 104, 55)",
                "fontSize": 0,
              },
            ]
          }
        >
          4
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "marginHorizontal": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "textAlign": "center",
            "width": 0,
          }
        }
      >
        Sangat Setuju
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "textAlign": "center",
            "width": 0,
          }
        }
      >
        Sangat Tidak Setuju
      </Text>
    </View>
  </View>
</View>
`;
